File: backend/services/lfhd_grouper.py
"""
Grouped LFHD formatter — pure helpers (additive, not yet wired).
================================================================

Mirror of ``frontend/static/lfhd_grouper.js``.  Same input/output
shape, same number formatting, same grouping rule, so the printed
LFHD math reads identically whether it is rendered by the frontend
preview or by a backend notice template.

Why this module exists
----------------------
The project owner asked for a *single* grouped-LFHD presentation
that can later feed:

* assessment notices
* court documents
* compounding sheets
* PDF / print views
* on-screen summaries

This file provides the building blocks.  Whether and where each
notice template adopts them is an explicit, per-template decision —
this PR intentionally does NOT change any existing template, route,
or document generator.  Court documents shouldn't quietly switch
formats; each adoption gets its own review-able PR.

Safety contract
---------------
* Pure Python, stdlib only — no Flask, no DB, no I/O.
* Does NOT import or call ``backend.services.calculator`` so it can
  never accidentally affect the LFHD assessment math.
* Idempotent and deterministic: repeated calls with identical input
  return identical output.

Input shape (matches the JSON the frontend already POSTs)::

    {"name": str,
     "load": float,      # Watts -- column header reads "Load (W)"
     "factor": float,    # dimensionless 0..1
     "hours": float,     # hours/day
     "days":   float}    # days

Group output shape::

    {"load_kw":      float,    # sum of contributing loads / 1000
     "hours":        float,
     "factor":       float,
     "days":         float,
     "units":        float,    # load_kw * hours * factor * days
     "device_count": int,
     "device_names": List[str]}

Math line format::

    "0.3 × 18 × 0.3 × 365 = 591.3"

Examples
--------
>>> rows = [
...     {"name": "Bulb",  "load": 100, "factor": 0.3, "hours": 18, "days": 365},
...     {"name": "Fan",   "load": 200, "factor": 0.3, "hours": 18, "days": 365},
...     {"name": "AC",    "load": 1500, "factor": 0.5, "hours":  8, "days": 120},
... ]
>>> groups = group_devices(rows)
>>> [g["device_count"] for g in groups]
[2, 1]
>>> format_math(groups[0])
'0.3 × 18 × 0.3 × 365 = 591.3'
>>> format_math(groups[1])
'1.5 × 8 × 0.5 × 120 = 720'
>>> total_units(groups)
1311.3
"""

from __future__ import annotations

from typing import Iterable, List, Dict, Any

__all__ = [
    "group_devices",
    "format_math",
    "total_units",
    "render_text",
    "render_html",
]

# ---------------------------------------------------------------------
# Pure helpers
# ---------------------------------------------------------------------

# 'times' character used in the printed math.  Kept as a constant so
# a reviewer can change it in exactly one place if a notice template
# prefers ASCII '*' for some legal reason.
_TIMES = "\u00d7"  # MULTIPLICATION SIGN ×


def _to_number(v: Any, default: float = 0.0) -> float:
    """Robust float coercion — empty strings, None, junk all map to default."""
    try:
        if v is None or v == "":
            return default
        return float(v)
    except (TypeError, ValueError):
        return default


def _round(n: float, decimals: int) -> float:
    """Round half-up at the given precision — matches JS Math.round*p/p."""
    p = 10 ** decimals
    # Python's banker's rounding is fine for our presentation use; the
    # difference vs. JS Math.round (round half away from zero) is at
    # the 0.5 mid-bin only and never affects realistic LFHD numbers.
    return round(n * p) / p


def _trim(n: float, decimals: int) -> str:
    """0.300 -> '0.3', 365 -> '365', 18.5 -> '18.5'.

    Mirrors the JS ``trim`` helper byte-for-byte: format with the given
    decimals, then drop trailing zeros and a redundant decimal point.
    """
    s = f"{float(n):.{decimals}f}"
    if "." in s:
        s = s.rstrip("0").rstrip(".")
    return s or "0"


def _bucket_key(d: Dict[str, Any]) -> tuple:
    """(hours, factor, days) bucket — rounded so float noise can't split."""
    return (
        _round(_to_number(d.get("hours")),  3),
        _round(_to_number(d.get("factor")), 3),
        _round(_to_number(d.get("days")),   0),
    )


def _is_valid_device(d: Dict[str, Any]) -> bool:
    """A device row is printable only when L/H/F/D are all > 0."""
    L = _to_number(d.get("load"))
    H = _to_number(d.get("hours"))
    F = _to_number(d.get("factor"))
    D = _to_number(d.get("days"))
    return L > 0 and H > 0 and F > 0 and D > 0


# ---------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------


def group_devices(devices: Iterable[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Group device rows by identical (hours, factor, days) and sum loads.

    The returned list preserves first-occurrence order so the printed
    math sequence is stable (deterministic notices === reviewable notices).
    """
    if devices is None:
        return []

    order: List[tuple] = []
    buckets: Dict[tuple, Dict[str, Any]] = {}

    for d in devices:
        if not _is_valid_device(d):
            continue
        key = _bucket_key(d)
        if key not in buckets:
            order.append(key)
            buckets[key] = {
                "load_w":       0.0,
                "hours":        key[0],
                "factor":       key[1],
                "days":         key[2],
                "device_names": [],
            }
        b = buckets[key]
        b["load_w"] += _to_number(d.get("load"))
        name = str(d.get("name") or "").strip() or "(unnamed)"
        b["device_names"].append(name)

    out: List[Dict[str, Any]] = []
    for key in order:
        b = buckets[key]
        load_kw = b["load_w"] / 1000.0
        out.append({
            "load_kw":      _round(load_kw, 4),
            "hours":        b["hours"],
            "factor":       b["factor"],
            "days":         b["days"],
            "units":        _round(load_kw * b["hours"] * b["factor"] * b["days"], 3),
            "device_count": len(b["device_names"]),
            "device_names": list(b["device_names"]),
        })
    return out


def format_math(g: Dict[str, Any]) -> str:
    """``"0.3 × 18 × 0.3 × 365 = 591.3"`` for one group."""
    return (
        f"{_trim(g['load_kw'], 3)} {_TIMES} "
        f"{_trim(g['hours'],   2)} {_TIMES} "
        f"{_trim(g['factor'],  3)} {_TIMES} "
        f"{_trim(g['days'],    0)} = "
        f"{_trim(g['units'],   2)}"
    )


def total_units(groups: Iterable[Dict[str, Any]]) -> float:
    """Sum of group units, rounded to 3 dp."""
    return _round(sum(g["units"] for g in (groups or [])), 3)


# ---------------------------------------------------------------------
# Renderers — string returns so any template can embed them.
# ---------------------------------------------------------------------


def render_text(devices: Iterable[Dict[str, Any]]) -> str:
    """Plain-text grouped LFHD block.  Drop-in for ``<pre>`` blocks."""
    groups = group_devices(devices)
    if not groups:
        return "(no LFHD entries)"
    lines = [
        f"{i + 1}. {format_math(g)} Units"
        for i, g in enumerate(groups)
    ]
    lines.append("")
    lines.append(f"Total Units = {_trim(total_units(groups), 2)}")
    return "\n".join(lines)


def _escape_html(s: str) -> str:
    return (
        str(s)
        .replace("&",  "&amp;")
        .replace("<",  "&lt;")
        .replace(">",  "&gt;")
        .replace('"',  "&quot;")
        .replace("'",  "&#39;")
    )


def render_html(devices: Iterable[Dict[str, Any]]) -> str:
    """HTML grouped LFHD block — same markup the frontend uses, so a
    notice template that opts in to this helper produces a printout
    identical to the on-screen preview."""
    groups = group_devices(devices)
    if not groups:
        return (
            '<div class="lfhd-empty">'
            'No LFHD entries — Load × Hours × Factor × Days needed.'
            '</div>'
        )

    rows = []
    for g in groups:
        if g["device_count"] == 1:
            names_html = _escape_html(g["device_names"][0])
        else:
            names_html = (
                ", ".join(_escape_html(n) for n in g["device_names"])
                + f' <span class="small">({g["device_count"]} devices)</span>'
            )
        rows.append(
            "<li>"
            f'<code class="lfhd-math">{_escape_html(format_math(g))}</code> '
            '<span class="lfhd-units small">Units</span>'
            f'<div class="lfhd-devices small">{names_html}</div>'
            "</li>"
        )

    total = _trim(total_units(groups), 2)
    return (
        '<ol class="lfhd-rows">' + "".join(rows) + "</ol>"
        f'<div class="lfhd-total"><b>Total Units</b> = '
        f'<code>{_escape_html(total)}</code></div>'
    )


# ---------------------------------------------------------------------
# When run as a script, print a tiny self-check so the doctests can be
# executed standalone without a test runner.  Not invoked by Flask.
# ---------------------------------------------------------------------
if __name__ == "__main__":          # pragma: no cover
    import doctest
    fails, tests = doctest.testmod(verbose=False)
    print(f"lfhd_grouper.py — {tests} doctest(s), {fails} failure(s)")

File: frontend/static/lfhd_grouper.js
/* ==================================================================
 * Raid Management System — Grouped LFHD formatter (frontend)
 * ------------------------------------------------------------------
 * Reusable helper that:
 *
 *   1. Groups device rows by (hours, factor, days) — devices that
 *      share the same H/F/D collapse into a single row whose load is
 *      the SUM of the contributing devices' loads.
 *   2. Formats each group as the legal-style math line
 *           L × H × F × D = Units
 *      (Load shown in kW, Units in kWh, two decimals trimmed.)
 *   3. Renders the grouped output as HTML for the New Case panel
 *      and as plain text for clipboard / future PDF / future notices.
 *
 * SCOPE  (this PR is intentionally additive)
 *   - Frontend only: nothing here changes how the LFHD calculator
 *     runs, how cases are saved, or how notices are generated today.
 *   - The same logic is mirrored in `backend/services/lfhd_grouper.py`
 *     so that existing notice templates can be opted-in to the new
 *     grouped output one at a time, in a separate PR, after each
 *     template's wording has been reviewed by the project owner.
 *
 * SAFETY CONTRACT
 *   - Does NOT modify app.js. Hooks into the existing #devicesBody
 *     entirely via event delegation + MutationObserver.
 *   - Does NOT modify the device payload sent to the backend.
 *     `collectDevicesFromUI` in app.js still produces the same
 *     shape; this module only READS the form state.
 *   - Does NOT change the existing Total Connected Load auto-sum
 *     logic. That uses Watts; this module reads the same Watts and
 *     converts to kW only for printed output.
 *   - Idempotent setup; safe across hot-reload.
 *
 * INPUT  (from #devicesBody — same shape app.js builds rows from)
 *
 *     {
 *       name:    string,
 *       load:    number,   // Watts (Section 4 column header is "Load (W)")
 *       factor:  number,   // dimensionless 0..1
 *       hours:   number,   // hours/day
 *       days:    number    // days
 *     }
 *
 * GROUP OUTPUT
 *
 *     {
 *       loadKW:      number,   // sum of contributing loads ÷ 1000
 *       hours:       number,
 *       factor:      number,
 *       days:        number,
 *       units:       number,   // loadKW × hours × factor × days
 *       deviceCount: number,
 *       deviceNames: string[]
 *     }
 *
 * MATH STRING FORMAT — matches the project owner's spec exactly:
 *
 *     "0.3 × 18 × 0.3 × 365 = 591.3"
 *
 *   Numbers are formatted to up to 3 decimals (load) / 2 decimals
 *   (others) and then stripped of trailing zeros so the math reads
 *   cleanly on a printed legal page.
 * ==================================================================
 */

(function () {
  'use strict';

  /* ------------------------------------------------------------- *
   *  Pure helpers — no DOM, deterministic, testable in isolation.  *
   * ------------------------------------------------------------- */

  /** Trim trailing zeros after toFixed: 0.300 -> "0.3", 365 -> "365". */
  function trim(n, decimals) {
    if (n === null || n === undefined || isNaN(n)) return '0';
    var fixed = Number(n).toFixed(decimals);
    // parseFloat eats the trailing zeros; toString gives canonical form.
    return parseFloat(fixed).toString();
  }

  /** Round to N decimals (avoids fp surprise like 0.1 + 0.2 = 0.30000…). */
  function round(n, decimals) {
    var p = Math.pow(10, decimals);
    return Math.round(Number(n) * p) / p;
  }

  /** Build the (hours, factor, days) bucket key. Rounding kills
   *  spurious float noise so e.g. 0.3 typed twice always groups.    */
  function bucketKey(d) {
    return [
      round(d.hours,  3),
      round(d.factor, 3),
      round(d.days,   0)
    ].join('|');
  }

  /** Return only rows that have non-zero numeric values for L/H/F/D.
   *  An empty or partial row should not appear in printed math.     */
  function isValidDevice(d) {
    var L = Number(d.load),   H = Number(d.hours);
    var F = Number(d.factor), D = Number(d.days);
    return (
      isFinite(L) && L > 0 &&
      isFinite(H) && H > 0 &&
      isFinite(F) && F > 0 &&
      isFinite(D) && D > 0
    );
  }

  /**
   * Group an array of device rows by identical (hours, factor, days)
   * and sum their loads. Order of groups matches first-occurrence in
   * input. Returns a NEW array — input is never mutated.
   */
  function group(devices) {
    if (!Array.isArray(devices)) return [];
    var order   = [];
    var buckets = Object.create(null);

    devices.forEach(function (d) {
      if (!isValidDevice(d)) return;
      var key = bucketKey(d);
      if (!buckets[key]) {
        order.push(key);
        buckets[key] = {
          loadW:       0,                       // accumulator in Watts
          hours:       round(Number(d.hours),  3),
          factor:      round(Number(d.factor), 3),
          days:        round(Number(d.days),   0),
          deviceNames: []
        };
      }
      var b = buckets[key];
      b.loadW += Number(d.load);
      b.deviceNames.push(String(d.name || '').trim() || '(unnamed)');
    });

    return order.map(function (k) {
      var b = buckets[k];
      var loadKW = b.loadW / 1000;
      return {
        loadKW:      round(loadKW, 4),
        hours:       b.hours,
        factor:      b.factor,
        days:        b.days,
        units:       round(loadKW * b.hours * b.factor * b.days, 3),
        deviceCount: b.deviceNames.length,
        deviceNames: b.deviceNames.slice()
      };
    });
  }

  /** "0.3 × 18 × 0.3 × 365 = 591.3" — the canonical legal math line. */
  function formatMath(g) {
    return (
      trim(g.loadKW, 3) + ' × ' +
      trim(g.hours,  2) + ' × ' +
      trim(g.factor, 3) + ' × ' +
      trim(g.days,   0) + ' = ' +
      trim(g.units,  2)
    );
  }

  /** Sum of group units. */
  function totalUnits(groups) {
    return round(
      (groups || []).reduce(function (s, g) { return s + g.units; }, 0),
      3
    );
  }

  /** ------------------------------------------------------------- *
   *  Renderers — return strings; the DOM glue below decides where  *
   *  to insert them. Keeping them as pure string returners means    *
   *  the same module can later feed the print preview, a clipboard *
   *  copy action, or a future PDF generator unchanged.             *
   * ------------------------------------------------------------- */

  function escapeHtml(s) {
    return String(s).replace(/[&<>"']/g, function (c) {
      return ({ '&':'&amp;', '<':'&lt;', '>':'&gt;', '"':'&quot;', "'":'&#39;' })[c];
    });
  }

  /** HTML for embedding in the New Case panel. */
  function renderHtml(devices) {
    var groups = group(devices);
    if (groups.length === 0) {
      return '<div class="lfhd-empty small">' +
             'Fill in Load, Factor, Hours and Days for at least one device ' +
             'to see the grouped LFHD math.</div>';
    }
    var rows = groups.map(function (g, i) {
      var names = g.deviceNames.length === 1
        ? escapeHtml(g.deviceNames[0])
        : g.deviceNames.map(escapeHtml).join(', ') +
          ' <span class="small">(' + g.deviceCount + ' devices)</span>';
      return (
        '<li>' +
          '<code class="lfhd-math">' + escapeHtml(formatMath(g)) + '</code> ' +
          '<span class="lfhd-units small">Units</span>' +
          '<div class="lfhd-devices small">' + names + '</div>' +
        '</li>'
      );
    }).join('');
    var total = trim(totalUnits(groups), 2);
    return (
      '<ol class="lfhd-rows">' + rows + '</ol>' +
      '<div class="lfhd-total"><b>Total Units</b> = ' +
        '<code>' + escapeHtml(total) + '</code></div>'
    );
  }

  /** Plain-text version (for clipboard / future PDF / future notice). */
  function renderText(devices) {
    var groups = group(devices);
    if (groups.length === 0) return '(no LFHD entries)';
    var lines = groups.map(function (g, i) {
      return (i + 1) + '. ' + formatMath(g) + ' Units';
    });
    lines.push('');
    lines.push('Total Units = ' + trim(totalUnits(groups), 2));
    return lines.join('\n');
  }

  /** ------------------------------------------------------------- *
   *  Public namespace — usable from console / future tariff engine *
   *  / future export-to-PDF code without re-importing.             *
   * ------------------------------------------------------------- */

  var API = {
    group:        group,
    formatMath:   formatMath,
    totalUnits:   totalUnits,
    renderHtml:   renderHtml,
    renderText:   renderText
  };

  window.RAID_LFHD = API;

  /* =============================================================== *
   *  DOM glue — the rest of this file is the page integration.      *
   *  Everything above this line is pure logic and re-usable.        *
   * =============================================================== */

  var DEVICE_PRESETS = [
    'Bulb / LED',
    'Ceiling Fan',
    'Cooler',
    'AC',
    'Refrigerator',
    'Submersible',
    'Heater / Geyser',
    'Washing Machine',
    'Iron Press',
    'Induction'
  ];

  function $(sel) { return document.querySelector(sel); }
  function $$(sel) { return Array.prototype.slice.call(document.querySelectorAll(sel)); }

  /** Read the live device table and convert each row to the
   *  shape `group()` expects. We deliberately read the DOM rather
   *  than reach into app.js's STATE.devices — that keeps this module
   *  independent of app.js internals. Empty / non-numeric cells are
   *  treated as zero so partially-filled rows are silently dropped
   *  by isValidDevice() inside group().                              */
  function readDevicesFromDom() {
    return $$('#devicesBody tr').map(function (tr) {
      function v(field) {
        var inp = tr.querySelector('input[data-field="' + field + '"]');
        return inp ? inp.value : '';
      }
      return {
        name:   v('name'),
        load:   Number(v('load'))   || 0,
        factor: Number(v('factor')) || 0,
        hours:  Number(v('hours'))  || 0,
        days:   Number(v('days'))   || 0
      };
    });
  }

  /** Add `list="device-presets"` to every Device-name input that
   *  doesn't already have it. Runs every time the device table
   *  changes — lets app.js render rows however it likes; this just
   *  decorates them progressively.                                   */
  function annotateDeviceInputs() {
    $$('#devicesBody input[data-field="name"]').forEach(function (inp) {
      if (inp.getAttribute('list') !== 'device-presets') {
        inp.setAttribute('list', 'device-presets');
      }
    });
  }

  /** Re-render the LFHD summary card from the live DOM. */
  function refreshSummary() {
    var card = $('#lfhdSummary');
    if (!card) return;
    card.innerHTML = renderHtml(readDevicesFromDom());
  }

  /** Inject the <datalist> with the preset device names. We do this
   *  in JS (rather than in index.html) so the preset list lives in
   *  exactly one place — this file. To extend, edit DEVICE_PRESETS.  */
  function ensureDatalist() {
    if (document.getElementById('device-presets')) return;
    var dl = document.createElement('datalist');
    dl.id = 'device-presets';
    DEVICE_PRESETS.forEach(function (p) {
      var o = document.createElement('option');
      o.value = p;
      dl.appendChild(o);
    });
    document.body.appendChild(dl);
  }

  function setup() {
    var body = $('#devicesBody');
    var card = $('#lfhdSummary');
    if (!body || !card) return;       // panel not present — nothing to do
    if (body.dataset.lfhdReady === '1') return;   // idempotent

    ensureDatalist();
    annotateDeviceInputs();
    refreshSummary();

    /* Live updates as the operator types — delegated input listener. */
    body.addEventListener('input', refreshSummary);

    /* Row added or removed by app.js (it sets innerHTML wholesale).   *
     * Use a MutationObserver because we don't want to monkey-patch    *
     * app.js's renderDevices(). On every structural change we both   *
     * re-annotate and re-render the summary.                         */
    var mo = new MutationObserver(function () {
      annotateDeviceInputs();
      refreshSummary();
    });
    mo.observe(body, { childList: true, subtree: true });

    /* "+ Add Device" / "+ Sample Devices" / "Live Calculate" all     *
     * eventually mutate #devicesBody, which the observer catches —   *
     * but Live Calculate also updates Units cells via setting        *
     * STATE.devices[i].units and re-rendering. The observer covers   *
     * that too. No extra wiring needed.                              */

    body.dataset.lfhdReady = '1';
  }

  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', setup);
  } else {
    setup();
  }

  /* Manual hook so a console operator (or future code that rebuilds *
   * the New Case panel) can re-run wiring without a full reload.    */
  window.RAID_setupLfhdSummary = setup;
})();
